### PR TITLE
add stream shutdown and support half-duplex operation

### PR DESCRIPTION
**Pull request by vtjnash**
Fixes one of the issues mentioned in #24526

### CODE REVIEW

**Comment by vtjnash** [comment]
CI failures are waiting for https://github.com/libuv/libuv/pull/3006

File: base/io.jl [hunk 1]
```diff
@@ -60,9 +60,36 @@ function isopen end
...
 """
 function close end
+
+"""
+    shutdown(stream)
+
+Shutdown an I/O stream. Performs a [`flush`](@ref) first. Notify the other end that no
+more data can ever be written to the underlying file. This is not supported by all IO
+types.
+"""
```
Comment (notinaboat) [inline base/io.jl:70]: Suggested wording tweak:
"""
Shutdown _**the write half of a full-duplex**_ I/O stream. Performs a flush first. Notify the other end that no more data _**will be written to the stream**_. This is not supported by all IO types.
"""
Perhaps include a pseudo-code example?
```
write(io, request)
shutdown(io)
result = read(io)
```

**Comment by c42f** [comment]
I like everything about this PR except for the newly exported name `shutdown()`.

Do we really want to export such a generic verb with such a relatively obscure meaning?  (Yes, it maps to the names used in the underlying system calls, but IMO that's the only thing it's got going for it.)

In #24526 it was suggested that `closewrite()` might be used which seems far more obvious.

**Comment by vtjnash** [comment]
Ah, I didn't see discussion of names. I am okay with it, since it dispatches on being an IO object

**Comment by c42f** [comment]
> I am okay with it, since it dispatches on being an IO object

To me this seems like `close()` but for the writer side? So `closewrite()` is a completely self-evident name if someone knows what `close()` does?  Whereas `shutdown()` has no analogy, except for the small number of Juila programmers who also know about `man 2 shutdown`.

The fact that it only works for `IO` objects is kind of the problem — other libraries should not really overload `Base.shutdown()` for unrelated functionality.

Is `closewrite()` a correct renaming? I'm happy to make a PR.

**Comment by vtjnash** [comment]
Yeah, there is also a `closeread` option, but generally seems pretty useless, and libuv doesn't expose it.

**Comment by notinaboat** [comment]
`closeread` is used in HTTP.jl: https://github.com/JuliaWeb/HTTP.jl/blob/7e4420d0733c87b3664657ac22d1bdc2fb86d789/src/IOExtras.jl#L1-L7

**Comment by c42f** [comment]
True. In HTTP, `closeread(::Transaction)` indicates that a transaction is done reading which unblocks the next transaction and enables it to start using the underlying TCP socket.

**Pull request merged**